src/main/java/org/sciborgs1155/robot/commands/NoteVisualizer.java

```diff
@@ -0,0 +1,190 @@
+              RED_MID_NOTE,
+              RED_RIGHT_NOTE));
+  private static boolean carryingNote = false;
+  private static ArrayList<Pose3d> pathPosition = new ArrayList<>();
+
+  // suppliers
+  private static Supplier<Pose2d> pose = Pose2d::new;
+  private static Supplier<Rotation3d> angle = Rotation3d::new;
+  private static DoubleSupplier velocity = () -> 1;
+
+  private static double zVelocity;
+  private static int i = 0;
+
+  private static Pose3d currentNotePose = new Pose3d();
+  private static Pose3d lastNotePose = new Pose3d();
+
+  // publishers
+  private static StructPublisher<Pose3d> shotNotePub;
+  private static StructArrayPublisher<Pose3d> pathPub;
+  private static StructArrayPublisher<Pose3d> notesPub;
+  private static BooleanPublisher hasNotePub;
+
+  public static void setSuppliers(
+      Supplier<Pose2d> robotPose, Supplier<Rotation3d> pivotAngle, DoubleSupplier shotVelocity) {
+    pose = robotPose;
+    angle = pivotAngle;
+    velocity = shotVelocity;
+  }
+
+  /** Set up NT publisher. Call only once before beginning to log notes. */
+  public static void startPublishing() {
+    NetworkTableInstance inst = NetworkTableInstance.getDefault();
+
+    StructArrayTopic<Pose3d> notesTopic = inst.getStructArrayTopic("notes", Pose3d.struct);
+    StructArrayTopic<Pose3d> notePathTopic = inst.getStructArrayTopic("note path", Pose3d.struct);
+    BooleanTopic hasNoteTopic = inst.getBooleanTopic("holding note");
+    StructTopic<Pose3d> shotNoteTopic = inst.getStructTopic("shot note", Pose3d.struct);
+
+    notesPub = notesTopic.publish();
+    pathPub = notePathTopic.publish();
+    hasNotePub = hasNoteTopic.publish();
+    shotNotePub = shotNoteTopic.publish();
+
+    notesPub.set(notes.toArray(new Pose3d[0]));
+  }
+
+  public static void log() {
+    shotNotePub.set(currentNotePose);
+    hasNotePub.set(carryingNote);
+  }
+
+  public static Command intake() {
+    if (carryingNote) {
+      return Commands.none();
+    }
+    return new ScheduleCommand(
+        Commands.defer(
+            () -> {
+              return Commands.run(
+                      () -> {
+                        Pose2d intakePose = pose.get();
+                        for (Pose3d note : notes) {
+                          double distance =
+                              Math.abs(
+                                  note.getTranslation()
+                                      .toTranslation2d()
+                                      .getDistance(intakePose.getTranslation()));
+                          if (distance > 0.2) {
+                            continue;
+                          }
+                          carryingNote = true;
+                          notes.remove(note);
+                          notesPub.set(notes.toArray(new Pose3d[0]));
+                          break;
+                        }
+                      })
+                  .until(() -> carryingNote);
+            },
+            Set.of()));
```

Comment:
add an `unless` clause at the end of these commands so that we only simulate intaking and shooting in sim